src/attributes.rs

```diff
@@ -0,0 +1,282 @@
+}
+
+impl<'a> Attribute<'a> {
+    pub fn as_string(&self) -> String {
+        format!(
+            "#{}[{}]",
+            if self.is_inner { "!" } else { "" },
+            self.content.raw_item
+        )
+    }
+
+    pub fn new(raw: &'a str) -> Self {
+        let raw_trimmed = raw.trim();
+        let raw_without_closing = raw_trimmed.strip_suffix(']').unwrap_or_else(|| {
+            panic!(
+                concat!(
+                    "String `{}` cannot be parsed as an attribute ",
+                    "because it is not closed with a square bracket."
+                ),
+                raw_trimmed
+            )
+        });
+
+        if let Some(raw_content) = raw_without_closing.strip_prefix("#[") {
+            Attribute {
+                is_inner: false,
+                content: Rc::new(AttributeMetaItem::new(raw_content)),
+            }
+        } else if let Some(raw_content) = raw_without_closing.strip_prefix("#![") {
+            Attribute {
+                is_inner: true,
+                content: Rc::new(AttributeMetaItem::new(raw_content)),
+            }
+        } else {
+            panic!(
+                concat!(
+                    "String `{}` cannot be parsed as an attribute ",
+                    "because it starts with neither `#[` nor `#![`."
+                ),
+                raw_trimmed
+            )
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct AttributeMetaItem<'a> {
+    pub raw_item: &'a str,
+    pub base: &'a str,
+    pub assigned_item: Option<&'a str>,
+    pub arguments: Option<Vec<Rc<AttributeMetaItem<'a>>>>,
+}
+
+impl<'a> AttributeMetaItem<'a> {
+    /// Tries to parse `raw` as a comma-separated sequence of `AttributeMetaItem`'s
+    /// wrapped in parentheses, square brackets or curly brackets.
+    fn slice_arguments(raw: &'a str) -> Option<Vec<Rc<AttributeMetaItem<'a>>>> {
+        let is_left = |c| c == '(' || c == '[' || c == '{';
+        let is_right = |c| c == ')' || c == ']' || c == '}';
+        let matching_right = |c| match c {
+            '(' => ')',
+            '[' => ']',
+            '{' => '}',
+            _ => unreachable!("Tried to find matching right bracket for {}.", c),
+        };
+
+        let raw_trimmed = raw.trim();
+        let first_char = raw_trimmed.chars().next()?;
+        let raw_meta_seq = raw_trimmed
+            .strip_prefix(is_left)?
+            .strip_suffix(|c| c == matching_right(first_char))?
+            .trim();
+
+        let mut i = 0; // index of the first character after the last comma
+        let mut bracket_depth = 0; // number of currently opened brackets
+        let mut arguments: Vec<Rc<AttributeMetaItem>> = Vec::new(); // meta items constructed so far
+
+        for (j, c) in raw_meta_seq.chars().enumerate() {
+            if is_left(c) {
+                bracket_depth += 1;
+            } else if is_right(c) {
+                bracket_depth -= 1;
+            } else if c == ',' {
```

Comment:
Can you add a test case that includes some parens and square brackets in a string literal, to make sure this parsing logic works correctly? I'm a bit worried about things like `#[foo("parens ) and brackets ] and curlies } here")]` being mis-parsed. Also similarly with open ones, or a mix of open and closed, and also with escaped quotes added in.

Reply:
This is certainly a problem, I'll start from fixing it.

Reply:
Should be fixed now!